Fix: Disable capa buttons to protect against race condition that may occur between two competing AJAX calls.

Qubad786 opened the pull request:
TNL-3477

**Background**: Someone clicked the Save button of capa problem, then shortly after clicked the Check button, and on the server side, the check completed before the save and the expected outcome was overwritten.  

**Fix**: Implemented disabling all capa buttons when reset, save or check is clicked so that competing AJAX calls cannot cause race condition to occur. Test cases are there to make sure, things work as intended.

@waheedahmed @adampalay, suggestions are always welcome!

waheedahmed commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,6 +703,42 @@ class @Problem
+    else
+      @showButton.addClass 'is-disabled'
+      @showButton.attr({'aria-disabled': 'true'})
+
```
I think it can be done by only one functio

adampalay replied:
Completely agree. Think about other ways you can make this code more DRY. Can you make a function, `disableButtonsWhileRunning`, that takes functions like `save_internal` or `reset` and disables all buttons while they're running?

Another thing to look out for: `enableCheckButton` changes the text of the check button while it's being clicked. We only want that when it's the check button that has been clicked. If you click "save", you don't want the check button's text to be changed, but you still want the check button to be disabled.

adampalay commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -198,6 +198,34 @@ describe 'Problem', ->
+          promise =
+            always: (callable) -> callable()
+            done: (callable) -> callable()
+        spyOn @problem, 'enableCheckButton'
```
can you make these tests more DRY?

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -32,11 +33,15 @@ class @Problem
     @checkButtonCheckingText = @checkButton.data('checking')
     @checkButton.click @check_fd
+    @hintButton = @$('div.action button.hint-button')
+    @hintButton.click @hint_button
```
i'm not familiar with this syntax — does this define what to do when `@hintButton` is clicked?

Qubad786 replied:
Yes.

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,16 +680,37 @@ class @Problem
       element = $(element)
       element.find("section[id^='forinput']").removeClass('choicetextgroup_show_correct')
 
+  disableAllButtonsWhileRunning: (buttonsHandlerCallback ,processCallback) =>
```
This function is unnessarily abstracted. `buttonsHandlerCallback` will only ever be `enableAllButtons`, so I think this function only needs to take the parameter `processCallback`.

By the way, why are you calling it `processCallback`? Isn't `operation` more descriptive, like how https://github.com/edx/edx-platform/blob/master/common/static/common/js/components/utils/view_utils.js#L116-L122 does it?

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,16 +680,37 @@ class @Problem
+      @hintButton.addClass('is-disabled').attr({'aria-disabled': 'true'})
+      @showButton.addClass('is-disabled').attr({'aria-disabled': 'true'})
+
   enableCheckButton: (enable) =>
```
`@change_check_button_text` should be a parameter in this function. Also, use camelCase for coffeescript.

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,16 +680,37 @@ class @Problem
+    # Used to enable/disable all buttons in problem.
+    @enableCheckButton enable
+    if enable
+      @resetButton.removeClass('is-disabled').attr({'aria-disabled': 'false'})
```
if you wanted to be super-dry, you could create one selector out of these by using [add](https://api.jquery.com/add/), so you'd only have to do `removeClass('is-disabled').attr({'aria-disabled': 'false'})` once

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,16 +676,48 @@ class @Problem
 
-  enableCheckButton: (enable) =>
+  disableAllButtonsWhileRunning: (operationCallback, isFromCheckOperation) =>
+    # Used to keep the buttons disabled while operationCallback is running.
```
add a comment to the code about why you need this `isFromCheckOperation` param

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,16 +676,48 @@ class @Problem
+
+  enableAllButtons: (enable, isFromCheckOperation) =>
+    # Used to enable/disable all buttons in problem.
+    if !isFromCheckOperation
```
I think what you want here is to put these @enableCheckButton calls in the if/else between L694-707. So something like:

```coffee
if enable
  # code here to remove the "is-disabled" class and aria attr
  @enableCheckButton true isFromCheckOperation

else
  # code here to *add* the "is-disabled" class and aria attr
  @enableCheckButton false isFromCheckOperation
```

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,16 +676,48 @@ class @Problem
+    operationCallback().always =>
+      @enableAllButtons true, isFromCheckOperation
+      if isFromCheckOperation
+        @enableCheckButtonAfterResponse
```
I don't think this is necessary anymore if you take my suggestion below

Qubad786 replied:
I'd like to tell you something before I remove @enableCheckButtonAfterResponse.

adampalay replied:
this logic should be moved into `enableAllButtons` below

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,16 +676,48 @@ class @Problem
+        @enableCheckButtonAfterResponse
+
+  enableAllButtons: (enable, isFromCheckOperation) =>
+    # Used to enable/disable all buttons in problem.
```
can you also add a comment in the code of what the parameters are?

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,16 +676,48 @@ class @Problem
+        .attr({'aria-disabled': 'true'})
+
+  enableCheckButton: (enable, changeText = true) =>
     # Used to disable check button to reduce chance of accidental double-submissions.
```
explain what `changeText` is in a code comment?

adampalay commented:
Tests look good, but have a few more comments about the code. It looks like you'll need a rebase too!

adampalay commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -676,16 +676,60 @@ class @Problem
+    #   'enable' is a boolean to determine enabling/disabling of buttons.
+    #   'isFromCheckOperation' is a boolean to keep track if operation was initiated
+    #    from @check.
+    if !isFromCheckOperation
```
I think my comment got lost in a rebase or something — these should be moved to the if/else in L702 - 717

Qubad786 replied:
Yeah seems , I didn't notice that one.

Qubad786 commented:
@adampalay, Have a look please.

adampalay commented:
:+1: 

@cahrens , may you please have a look?

@caesar2164 and @stvstnfrd , this is a generalization of the work you did to gray out the check button when clicking "save". Now, whenever you click on one of these buttons that can update CSM, each of the other buttons are grayed out to avoid race conditions.

stvstnfrd commented:
Thanks for the heads-up and thanks for the followup work here!

adampalay commented:
@stvstnfrd no problem! I'll give you a shoutout when we merge this in too :)

cahrens commented:
Sorry for the slow code review response... I've been out of the office with sick kiddos.

Is this branch rebased on master (following the django 1.8 upgrade)? I'd like to check out the branch and test locally when I review.

Qubad786 commented:
@cahrens yes.

cahrens commented:
Thank you, I should be able to review this today.

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -674,16 +676,58 @@ class @Problem
+    # Used to enable/disable all buttons in problem.
+    # params:
+    #   'enable' is a boolean to determine enabling/disabling of buttons.
+    #   'isFromCheckOperation' is a boolean to keep track if operation was initiated
```
It would be helpful to extend this documentation to explain why it is necessary to know if the operation was initiated from check. From what I can tell, it mostly has to do with changing the text on the check button.

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -198,17 +198,39 @@ describe 'Problem', ->
+            done: (callable) -> callable()
+        spyOn @problem, 'enableAllButtons'
+        @problem.check()
+        expect(@problem.enableAllButtons).toHaveBeenCalledWith false, true
```
Can you add coverage for the expected change in text on the check button?

cahrens commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.coffee:
```diff
@@ -216,9 +238,27 @@ describe 'Problem', ->
+            always: (callable) -> callable()
+        spyOn @problem, 'enableAllButtons'
+        @problem.reset()
+        expect(@problem.enableAllButtons).toHaveBeenCalledWith false, false
```
Can you verify that the text does not change on the check button?

cahrens commented on common/lib/xmodule/xmodule/js/src/capa/display.coffee:
```diff
@@ -322,9 +322,11 @@ class @Problem
-    ).always(@enableCheckButtonAfterResponse)
 
   reset: =>
+    @disableAllButtonsWhileRunning @reset_internal, false
```
So "show" and "hint" don't need to disable buttons?

cahrens commented:
If an error is returned from "Check", the Check button never gets re-enabled.

![image](https://cloud.githubusercontent.com/assets/484484/11512676/3d2d65aa-983d-11e5-8fed-875576169baa.png)

cahrens commented:
I checked out the branch and did some manual testing (finding the bug above). For the next round of review, can you please link a sandbox with a course that has a problem with all 5 buttons present? I'm not sure how to show the "hint" and"save" buttons. Thanks!

Qubad786 commented:
@cahrens [here](https://disable-capa-buttons.sandbox.edx.org/courses/course-v1:Arbisoft+CS101+2014_LT/courseware/21d049803c5f4300936cec486a67e181/a23962f3e74247feba3cc4016a72de55/1?activate_block_id=block-v1%3AArbisoft%2BCS101%2B2014_LT%2Btype%40vertical%2Bblock%408e4db75f7ed448f083fb1af8f7b16d06) is the sandbox with couple of  different problem components with required buttons. 

What I get from the code is, 'show answer' and 'hint' buttons do not affect the state of the problem (so cannot cause race condition) thats why they don't need to disable rest of the buttons. The bug was due to a function whose instance was simply returned instead of being called when coffee script code was compiled to javascript code as I didn't use parentheses. Anyways, I have refactored it and added the math input problems to sandbox for you to check as well.

Also, I have extended the documentation to provide more information about the extra parameter passed as an argument to methods and added the tests for expected change in text of check button and to verify that the text does not change on the check button on clicking any of reset, save buttons. Further suggestions are welcome! 

Thanks.

cahrens commented:
:+1: I did some quick testing on the sandbox and skimmed through the PR. It was difficult for me to tell though what changed since the last review, since you squashed the commits into a single commit. In the future, please don't squash review feedback commits until the reviewer has had a chance to look them over.

Qubad786 commented:
Oh really sorry, I didn't realize that one. I will be making sure of this in future.

adampalay commented:
:+1:
